// .ignored/next/dist/esm/lib/metadata/default-metadata.js
import React from "react";
export const createDefaultMetadata = ()=>{
    return {
        viewport: "width=device-width, initial-scale=1",
        // Other values are all null
        metadataBase: null,
        title: null,
        description: null,
        applicationName: null,
        authors: null,
        generator: null,
        keywords: null,
        referrer: null,
        themeColor: null,
        colorScheme: null,
        creator: null,
        publisher: null,
        robots: null,
        manifest: null,
        alternates: {
            canonical: null,
            languages: null,
            media: null,
            types: null
        },
        icons: null,
        openGraph: null,
        twitter: null,
        verification: {},
        appleWebApp: null,
        formatDetection: null,
        itunes: null,
        abstract: null,
        appLinks: null,
        archives: null,
        assets: null,
        bookmarks: null,
        category: null,
        classification: null,
        other: {}
    };
};
export const DEFAULT_METADATA_TAGS = [
    /*#__PURE__*/ React.createElement("meta", {
        charSet: "utf-8",
        key: "charset"
    }),
    /*#__PURE__*/ React.createElement("meta", {
        name: "viewport",
        content: "width=device-width, initial-scale=1",
        key: "viewport"
    }), 
];
