// .ignored/next/dist/lib/metadata/default-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DEFAULT_METADATA_TAGS = exports.createDefaultMetadata = void 0;
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const createDefaultMetadata = ()=>{
    return {
        viewport: "width=device-width, initial-scale=1",
        // Other values are all null
        metadataBase: null,
        title: null,
        description: null,
        applicationName: null,
        authors: null,
        generator: null,
        keywords: null,
        referrer: null,
        themeColor: null,
        colorScheme: null,
        creator: null,
        publisher: null,
        robots: null,
        manifest: null,
        alternates: {
            canonical: null,
            languages: null,
            media: null,
            types: null
        },
        icons: null,
        openGraph: null,
        twitter: null,
        verification: {},
        appleWebApp: null,
        formatDetection: null,
        itunes: null,
        abstract: null,
        appLinks: null,
        archives: null,
        assets: null,
        bookmarks: null,
        category: null,
        classification: null,
        other: {}
    };
};
exports.createDefaultMetadata = createDefaultMetadata;
const DEFAULT_METADATA_TAGS = [
    /*#__PURE__*/ _react.default.createElement("meta", {
        charSet: "utf-8",
        key: "charset"
    }),
    /*#__PURE__*/ _react.default.createElement("meta", {
        name: "viewport",
        content: "width=device-width, initial-scale=1",
        key: "viewport"
    }), 
];
exports.DEFAULT_METADATA_TAGS = DEFAULT_METADATA_TAGS;
